src/llmcompressor/modeling/moe_context.py

```diff
@@ -0,0 +1,177 @@
+    ) -> "MoECalibrationModule":
+        """
+        Create a calibration module from the original MoE module.
+
+        Args:
+            original: The original MoE module to convert
+            config: Model configuration (contains num_experts, etc.)
+            calibrate_all_experts: If True, send all tokens to all experts.
+                                   If False, use normal routing.
+
+        Returns:
+            Instance of the calibration module
+        """
+        pass
+
+    def restore(self) -> torch.nn.Module:
+        """
+        Restore the original module structure.
+
+        Only needed if is_permanent=False. For permanent modules, this is a no-op.
+
+        Returns:
+            The original module (or self if permanent)
+        """
+        if self.is_permanent:
+            return self
+        raise NotImplementedError(
+            f"{self.__class__.__name__} has is_permanent=False but doesn't "
+            "implement restore()"
+        )
+
+
+# Registry: module class name -> calibration module class
+MOE_CALIBRATION_MODULES: Dict[str, Type[MoECalibrationModule]] = {}
+
+
+def register_moe_calibration(module_class_name: str):
+    """
+    Decorator to register a MoE calibration module.
+
+    Usage:
+        @register_moe_calibration("DeepseekV3MoE")
+        class CalibrationDeepseekV3MoE(MoECalibrationModule):
+            ...
+
+    Args:
+        module_class_name: The class name of the original module to replace
+    """
+
+    def decorator(cls: Type[MoECalibrationModule]) -> Type[MoECalibrationModule]:
+        if not issubclass(cls, MoECalibrationModule):
+            raise TypeError(f"{cls.__name__} must inherit from MoECalibrationModule")
+        MOE_CALIBRATION_MODULES[module_class_name] = cls
+        return cls
+
+    return decorator
+
+
+@contextlib.contextmanager
+def moe_calibration_context(
+    model: PreTrainedModel,
+    calibrate_all_experts: bool = True,
+):
+    """
+    Context manager that applies MoE calibration to a model.
+
+    This scans all modules in the model and replaces any MoE modules with their
+    calibration equivalents. After the context exits, non-permanent modules are
+    restored to their original form.
+
+    The model is modified in-place, so the same model object should be used
+    within the context.
+
+    Args:
+        model: The model to apply MoE calibration to (modified in-place)
+        calibrate_all_experts: If True, all experts see all tokens during calibration.
+                               If False, use normal routing (useful for some techniques)
+
+    Example:
+        with moe_calibration_context(model):
+            # Run calibration - all experts will see data
+            for batch in dataloader:
+                model(**batch)
+        # Model is now restored (unless permanent)
+    """
+    replaced = {}
+
+    # Step 1: Find and replace MoE modules
+    for name, module in model.named_modules():
```

Comment:
Iterating through this might take some time, especially if there are modules to replace. You'll want to make sure the user has some feedback, rather than hanging right before calibration starts without explanation.

Consider doing one pass to collect all the `(name, module)`s to replace, then a tqdm iteration to replace those modules, along with any logging.